pubsub/natspubsub/nats.go

```diff
@@ -39,6 +54,95 @@ import (
 
 var errNotInitialized = errors.New("natspubsub: topic not initialized")
 
+func init() {
+	o := new(defaultDialer)
+	pubsub.DefaultURLMux().RegisterTopic(Scheme, o)
+	pubsub.DefaultURLMux().RegisterSubscription(Scheme, o)
+}
+
+// defaultDialer dials a default NATS server based on the environment
+// variable "NATS_SERVER_URL".
+type defaultDialer struct {
+	init   sync.Once
+	opener *URLOpener
+	err    error
+}
+
+func (o *defaultDialer) defaultConn(ctx context.Context) (*URLOpener, error) {
+	o.init.Do(func() {
+		serverURL := os.Getenv("NATS_SERVER_URL")
+		if serverURL == "" {
+			o.err = errors.New("NATS_SERVER_URL environment variable not set")
+			return
+		}
+		conn, err := nats.Connect(serverURL)
+		if err != nil {
+			o.err = fmt.Errorf("failed to dial NATS_SERVER_URL %q: %v", serverURL, err)
+			return
+		}
+		o.opener = &URLOpener{Connection: conn}
+	})
+	return o.opener, o.err
+}
+
+func (o *defaultDialer) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic, error) {
+	opener, err := o.defaultConn(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("open topic %v: failed to open default connection: %v", u, err)
+	}
+	return opener.OpenTopicURL(ctx, u)
+}
+
+func (o *defaultDialer) OpenSubscriptionURL(ctx context.Context, u *url.URL) (*pubsub.Subscription, error) {
+	opener, err := o.defaultConn(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("open subscription %v: failed to open default connection: %v", u, err)
+	}
+	return opener.OpenSubscriptionURL(ctx, u)
+}
+
+// Scheme is the URL scheme natspubsub registers its URLOpeners under on pubsub.DefaultMux.
+const Scheme = "nats"
+
+// URLOpener opens NATS URLs like "nats://mytopic".
+//
+// The URL host+path is used as the topic name.
+//
+// No query parameters are supported.
+type URLOpener struct {
+	// Connection to use for communication with the server.
+	Connection *nats.Conn
+	// TopicOptions specifies the options to pass to OpenTopic.
+	TopicOptions TopicOptions
+	// SubscriptionOptions specifies the options to pass to OpenSubscription.
+	SubscriptionOptions SubscriptionOptions
+}
+
+// OpenTopicURL opens a pubsub.Topic based on u.
+func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic, error) {
+	for param := range u.Query() {
+		return nil, fmt.Errorf("open topic %v: invalid query parameter %s", u, param)
+	}
+	topicName := path.Join(u.Host, u.Path)
+	return CreateTopic(o.Connection, topicName, &o.TopicOptions)
+}
+
+// OpenSubscriptionURL opens a pubsub.Subscription based on u.
+func (o *URLOpener) OpenSubscriptionURL(ctx context.Context, u *url.URL) (*pubsub.Subscription, error) {
+	for param := range u.Query() {
+		return nil, fmt.Errorf("open subscription %v: invalid query parameter %s", u, param)
+	}
+	topicName := path.Join(u.Host, u.Path)
+	return CreateSubscription(o.Connection, topicName, func() {}, &o.SubscriptionOptions)
```

Comment:
This isn't acceptable to me. Either the ack func should panic by default, or there should not be a default.

If you choose the latter, then I'd be OK with a required query param ackfunc=nop or ackfunc=panic. Although it's weird to have a required query param, isn't it?

Reply:
Defaulted to `panic`.